sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/tuplesketchesAggregates.scala

```diff
@@ -0,0 +1,843 @@
+      lgNomEntriesExpr = Some(newSecond),
+      summaryTypeExpr = newThird,
+      modeExpr = newFourth)
+
+  // Overrides for TypedImperativeAggregate
+
+  override def prettyName: String = "tuple_sketch_agg"
+
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(
+      StructType,
+      IntegerType,
+      StringTypeWithCollation(supportsTrimCollation = true),
+      StringTypeWithCollation(supportsTrimCollation = true))
+
+  override def dataType: DataType = BinaryType
+
+  override def nullable: Boolean = false
+
+  override def first: Expression = child
+  override def second: Expression =
+    lgNomEntriesExpr.getOrElse(Literal(ThetaSketchUtils.DEFAULT_LG_NOM_LONGS))
+  override def third: Expression = summaryTypeExpr
+  override def fourth: Expression = modeExpr
+
+  /**
+   * Extract and cache the key and summary value types from the input struct. Field 0 is the key
+   * type, Field 1 is the summary value type.
+   *
+   * Note: The asInstanceOf[StructType] cast is safe because inputTypes enforces that the first
+   * parameter must be StructType. This is validated during query analysis before execution.
+   */
+  private lazy val structType = child.dataType.asInstanceOf[StructType]
+  private lazy val keyType = structType.fields(0).dataType
+  private lazy val valueType = structType.fields(1).dataType
+
+  /**
+   * Factory for creating summary objects based on the input summary type and aggregation mode.
+   */
+  private lazy val summaryFactoryInput =
+    ThetaSketchUtils.getSummaryFactory(summaryTypeInput, modeInput)
+
+  /**
+   * Instantiate an UpdatableSketch instance using the lgNomEntries param and summary factory.
+   *
+   * @return
+   *   an UpdatableSketch instance wrapped with UpdatableTupleSketchBuffer
+   */
+  override def createAggregationBuffer(): TupleSketchState = {
+    val builder = new UpdatableSketchBuilder[Any, UpdatableSummary[Any]](summaryFactoryInput)
+    builder.setNominalEntries(1 << lgNomEntriesInput)
+    val sketch = builder.build()
+    UpdatableTupleSketchBuffer(sketch)
+  }
+
+  /**
+   * Evaluate the input row and update the UpdatableSketch instance with the row's key and summary
+   * value. The update function only supports a subset of Spark SQL types, and an exception will
+   * be thrown for unsupported types. Notes:
+   *   - Null values are ignored.
+   *   - Empty byte arrays are ignored
+   *   - Empty arrays of supported element types are ignored
+   *   - Strings that are collation-equal to the empty string are ignored.
+   *
+   * @param updateBuffer
+   *   A previously initialized UpdatableSketch instance
+   * @param input
+   *   An input row
+   */
+  override def update(updateBuffer: TupleSketchState, input: InternalRow): TupleSketchState = {
+    // Return early for null values.
+    val structValue = child.eval(input)
+    if (structValue == null) return updateBuffer
+
+    // Safe: child.eval() returns InternalRow when child.dataType is StructType
```

Comment:
Do we enforce this with `checkInputDataTypes` or somewhere else?

Reply:
yes, checkInputDataTypes will throw an error if it is not a struct type, example below:

```
val result = spark.sql("""
     |   SELECT tuple_sketch_agg(col, 15)
     |   FROM VALUES (50, 1.0D), (60, 2.0D), (60, 3.0D) tab(col, val)
     | """)
     | 
     | result.show()
org.apache.spark.sql.catalyst.ExtendedAnalysisException: [DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE] Cannot resolve "tuple_sketch_agg(col, 15, double, sum)" due to data type mismatch: The first parameter requires the "STRUCT" type, however "col" has the type "INT". SQLSTATE: 42K09; line 2 pos 9;
```